### ISSUE TITLE
[FEA] Groupby nth method

### ISSUE THREAD

**Issue by ayushdg** [rapidsai/cudf#1841]
**Is your feature request related to a problem? Please describe.**
I have workflow that utilizes the groupby `first` and `last` methods. It would generally add more flexibility if we add support for a nth method like pandas: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.GroupBy.nth.html

**Describe the solution you'd like**
Rather than having a separate implementation for `last` as a feature #1476 we could have an implementation for nth and just use nth(0) etc for `first` and `last`

**Describe alternatives you've considered**
Right now workarounds for this is to use the legacy groupby with a custom function

**Comment by randerzander** [comment]
@jrhemstad I take it this would only work for sort based groupbys?

@kkraus14 this would avoid a nasty workaround for @ayushdg 's workflow. Does this look at all feasible for 0.8, and is it blocked by #1747 ?

**Comment by kkraus14** [comment]
> @jrhemstad I take it this would only work for sort based groupbys?
> 
> @kkraus14 this would avoid a nasty workaround for @ayushdg 's workflow. Does this look at all feasible for 0.8, and is it blocked by #1747 ?

Yes this would only work for sort based groupbys, no it's not feasible for 0.8 unless you could have it work with the `cudf` method via a UDF currently.

**Comment by abellina** [comment]
In spark, there are two specific cases that have extra semantics. `first` and `last`. In both of these cases, one can have the option to ignore nulls or keep them. So a `first(null, 4, 5)` could return `null` or `4` depending on this flag.

**Comment by abellina** [comment]
One more point, in Spark we also have a `coalesce` operation. This would pick the first non-null element. It could be done a part of a group by, or standalone (`select coalesce(null, 2)` => would return 2).

**Comment by karthikeyann** [comment]
Here is a test code in pandas for different behavior of `groupby.nth`.
```python
import pandas as pd
import numpy as np
df = pd.DataFrame({'A': [0, 1, 1, 2, 1, 2],^M
                                'B': [np.nan, np.nan, 5, 4, 3, 2]}, columns=['A', 'B'])
g = df.groupby('A')
 for d in [None, 'any', 'all']:
             for n in [0,1]:
                 try:
                     print("n:", n, ",dropna:",str(d),",\n",g.nth(n,dropna=d))
                 except:
                     print("n:", n, ",dropna:",str(d),", ERROR")
```


output:
```
n: 0 ,dropna: None , (n exists in all groups)
      B
A
0  NaN
1  NaN
2  4.0
n: 1 ,dropna: None ,  (n does not exist in one of the groups -  that group key is skipped)
      B
A
1  5.0
2  2.0
n: 0 ,dropna: any ,  (n exists in all groups)
      B
A
0  NaN
1  5.0
2  4.0
n: 1 ,dropna: any , (n does not exist in one of the groups -  that group key output value is NAN)
      B
A
0  NaN
1  3.0
2  2.0
n: 0 ,dropna: all , (behaves similar to nth(0,dropna=None))
      B
A
0  NaN
1  NaN
2  4.0
n: 1 ,dropna: all , ERROR  (ValueError: Length mismatch: Expected axis has 2 elements, new values have 3 elements)
```

- nth with `dropna='any'` can be implemented with current groupby infrastructure.  
  - `groupby.first() == groupby.nth(0, dropna='any')`
  - `groupby.first() == groupby.nth(-1, dropna='any')`
- but nth with `dropna=None`, where n>sizeof group, cannot replicate exact pandas output because some output keys cannot be skipped. _What should be the libcudf++ nth behavior here?_
- nth with `dropna='all'` is similar to `dropna=None` after df.dropna('all') but instead of skipping, it creates error. 
  `dropped = self.obj.dropna(how=dropna, axis=self.axis)` and does nth as usual.
https://github.com/pandas-dev/pandas/blob/v1.0.1/pandas/core/groupby/groupby.py#L1804
this includes dropna on keys+values, we can't implement this with current infra).

Please give your suggestions. especially which are the features **required**.

**Comment by karthikeyann** [comment]
> In spark, there are two specific cases that have extra semantics. `first` and `last`. In both of these cases, one can have the option to ignore nulls or keep them. So a `first(null, 4, 5)` could return `null` or `4` depending on this flag.

@abellina what should be result of `first(null, null)` with ignore nulls?

**Comment by karthikeyann** [comment]
@kkraus14

**Comment by abellina** [comment]
@karthikeyann  in this case, the behavior is just to return null.

For a dataset like this:
```
scala> spark.sql("select * from (select value, null as all_nulls from (select explode(array(null, 1,2,3,4,5,6)) as value))").show
+-----+---------+
|value|all_nulls|
+-----+---------+
| null|     null|
|    1|     null|
|    2|     null|
|    3|     null|
|    4|     null|
|    5|     null|
|    6|     null|
+-----+---------+
```
For a column with valid values and nulls:
- `skip nulls = true` => skip nulls and produce a valid output.
- `skip nulls = false` => not skipping nulls, can produce a null.

For a column with all nulls:
- `skip nulls = true` => return null.
- `skip nulls = false` => return null.

```
scala> spark.sql("select first(value, false), first(value, true), first(all_nulls, false), first(all_nulls, true) from (select value, null as all_nulls from (select explode(array(null, 1,2,3,4,5,6)) as value))").show
+-------------------+------------------+-----------------------+----------------------+
|first(value, false)|first(value, true)|first(all_nulls, false)|first(all_nulls, true)|
+-------------------+------------------+-----------------------+----------------------+
|               null|                 1|                   null|                  null|
+-------------------+------------------+-----------------------+----------------------+
```

**Comment by kkraus14** [comment]
> but nth with dropna=None, where n>sizeof group, cannot replicate exact pandas output because some output keys cannot be skipped. What should be the libcudf++ nth behavior here?

It's not possible to return `null` in this situation?

> nth with dropna='all' is similar to dropna=None after df.dropna('all') but instead of skipping, it creates error.

I'm fine with returning `null` in this situation

**Comment by karthikeyann** [comment]
> > but nth with dropna=None, where n>sizeof group, cannot replicate exact pandas output because some output keys cannot be skipped. What should be the libcudf++ nth behavior here?
> 
> It's not possible to return `null` in this situation?

libcudf can return `null` in this situation.

**Comment by harrism** [comment]
Is Python plumbing still remaining for this issue?

**Comment by kkraus14** [comment]
I believe this is fixed by #4346
